rfc/rfc-0000-allow-resources-to-report-status.md

```diff
@@ -0,0 +1,320 @@
+metadata:
+  name: deploy
+spec:
+  resourceStatusRef:
+    name: kapp-rule
+
+  template:
+    apiVersion: kappctrl.k14s.io/v1alpha1
+    kind: App
+    metadata:
+      name: $(workload.metadata.name)$
+    spec:
+      ...
+
+---
+apiVersion: carto.run/v1alpha1
+kind: Workload
+metadata:
+  ...
+spec:
+  ...
+status:
+  conditions:
+    ...
+  resources:
+    - name: deployer
+      outputs:
+        ...
+      stampedRef:
+        ...
+      templateRef:
+        apiVersion: carto.run/v1alpha1
+        kind: ClusterTemplate
+        name: deploy
+      stampedStatus:
+        status: 'Failed'
+        fields:
+          - key: status.conditions['ReconcileFailed'].status
+            value: True
+            message: "Some kapp useful error message"
+```
+
+```yaml
+---
+apiVersion: carto.run/v1alpha1
+kind: ClusterResourceStatusRule
+metadata:
+  name: service-rule
+spec:
+  succeeded:
+    matchFields:
+      - key: '.status.loadBalancer'
+        operator: 'Exists'
+  failed:
+    matchFields:
+      - key: '.status.loadBalancer'
+        operator: 'DoesNotExist'
+
+---
+apiVersion: carto.run/v1alpha1
+kind: ClusterTemplate
+metadata:
+  name: servicer
+spec:
+  resourceStatusRef:
+    name: service-rule
+
+  template:
+    apiVersion: v1
+    kind: Service
+    metadata:
+      name: $(workload.metadata.name)$
+    spec:
+      ...
+
+---
+apiVersion: carto.run/v1alpha1
+kind: Workload
+metadata:
+  ...
+spec:
+  ...
+status:
+  conditions:
+    ...
+  resources:
+    - name: service-creator
+      outputs:
+        ...
+      stampedRef:
+        ...
+      templateRef:
+        apiVersion: carto.run/v1alpha1
+        kind: ClusterTemplate
+        name: servicer
+      stampedStatus:
+        status: 'Succeeded'
+        fields:
+          - key: status.loadBalancer
+            value: {}
+```
+
+# How it Works
+[how-it-works]: #how-it-works
+
+A stamped resource can be in one of three states:
+1. 'Succeeded'
+2. 'Failed'
+3. 'Unknown'
+
+It is up to each template author to determine what "success" and "failure" mean for the given resource the template is stamping out. "Unknown" strictly represents that Cartographer has not be able to determine success or failure.
+
+A `ClusterResourceStatusRule` requires defining matchers for both the `Succeeded` state and the `Failed` state.
+The matchers in the `Succeeded` state are ANDed together, whereas the matchers in the `Failed` state are ORed. That is, a resource is considered to have `Succeeded` only if _all of_ the matchers are fulfilled, but is considered to have `Failed` if _any one of_ the matchers are fulfilled.
+
+The supported matchers are `matchConditions` and `matchFields`. Each is an array, and an author can define either or both of these matchers for a given state.
+
+For `matchConditions`, authors need only define the `type` and `status` of the condition for Cartographer to read. In the owner status, the full condition will be reflected (with the `reason`, `message`, etc.)
```

Comment:
if the `matchConditions` array is populated with more than one item (say, 2), would we then bubble up to the owner status both of the conditions that matched?

Reply:
Yes

Reply:
If there is more than one, which is the canonical top level status. That is, looking at a particular resource in the workload, whats the source of truth for the status of the resource?

Reply:
We will bubble up anything the template author provides and that will be the source of truth. That is not for Cartographer to determine.

Reply:
In the in person design discussion, we were talking about being able to provide one entry to refer to (including programatically) for the state of an underlying (stamped) resource. I thought the idea here was to help developers rapidly identify where their issue is, including in visualisation tools?

Reply:
I think i'm just being slow. The source of truth is stampedStatus.status ?